---
title: "Analyses de données single-cell RNA-seq"
author: "Nathalie Lehmann & Lorette Noiret"
output: bookdown::gitbook
documentclass: book
bibliography: []
biblio-style: apalike
link-citations: yes
---

# Introdution au Chapitre 4 {.unnumbered}

## Objectifs {.unnumbered}

Dans ce dernier chapitre, vous allez apprendre à :

* Identifier des populations cellulaires d'intérêt grâce au **clustering** ;
* Analyser des gènes d'intérêt pour chacune de ces populations grâce aux **analyses d'expression différentielle**

Il s'agit... extraire du sens biologique...


## Données {.unnumbered}

Dans la continuité du chapitre 3, nous allons nous baser sur le package [Seurat](https://satijalab.org/seurat/articles/get_started.html) pour ces analyses.

```{r load-library, message = FALSE}
library(Seurat)
library(dplyr)
library(RColorBrewer)
library(ggplot2)
```

Nous gardons également les mêmes données que pour le chapitre 3. Pour rappel, les données sont issues de l'article  [@Moyer2020]. Si vous avez déjà un objet R (`rds` ou `Rda`) vous pouvez le charger directement pour poursuivre les analyses. Sinon, vous trouverez les données dans le dossier `data/r_objects/` du répertoire correspondant au chapitre.

Voici un exemple pour les charger avec la fonction `readRDS` :

```{r load-data}
sobj <- readRDS(file = "data/r_objects/SC26_FMTG_sobj_filtered_normalized.rds")
```

Pour vérifier le contenu de l'objet `sobj`, il suffit d'entrer le nom de l'objet dans la console :

```{r sobj}
sobj
```

L'output indique que nous avons bien chargé un fichier appelé `sobj` qui contient l'objet Seurat, de 31682 gènes et 2247 cellules.

A tout moment, vous pourrez sauvergarder vos fichiers avec la fonction `saveRDS` :

```{r save-data, eval = FALSE}
saveRDS(object = sobj,
        file = "data/r_objects/SC26_FMTG_sobj_saved.rds")
```


## Manipuler un objet Seurat (rappels) {.unnumbered}

A Seurat object is not the easiest structure to work with, but with a bit of practice you will learn to appreciate its potentiality.

In Seurat, data are organised in different compartements (slots), which contain themselves several compartements, which can also contain sub-compartments, etc.

Each compartment can be used to store:

* data from multiple modalities, such as RNAseq, spatial transcriptomics, ATAC-seq… For our session today, we will only focus on scRNAseq data (slot `assays`, sub-slot `RNA`)
* general results regarding your data, e.g. the total number of UMI expressed (slot `meta.data`)
* results of analyses: PCA components or clustering results

```{r slotnames, eval=FALSE}
slotNames(mydata)
```
You navigate through this hierarchy using `@` and `$` signs.
```{r slotnames-assays, eval=FALSE}
slotNames(mydata@assays$RNA)
```
In the slots associated RNA, you can store:

* `counts` : raw UMI (the data we imported)
* `data` : filtered/normalized counting matrix 
* `scale.data` : normalized and scaled data (usually for PCA analyses)
* `var.features`: contains a list of genes genes that contribute strongly to cell-to-cell variation (see section \@ref(hvg) on highly variable genes). 

You can access the data directly with the `GetAssayData` function.
```{r getassay, eval=FALSE}
# mydata@assays$RNA@counts[1:3,1:5]
GetAssayData(mydata, slot="counts")[1:3,1:5]
```
In Seurat, data are stored as "dgCMatrix", which is an efficient way to store an array with a lot of zeros in a computer (sparse matrix). 

```{r plots-filtered, eval=FALSE}
# the function "subset" helps you to filter the cells
# we create a new seurat object containing the filtred cells
mydata_filtrd <- subset(mydata, subset = nFeature_RNA > minGene &
						nFeature_RNA < maxGene &
						nCount_RNA > minUMI &
						nCount_RNA < maxUMI &
						percent.mt < maxpct_mt)

VlnPlot(mydata_filtrd, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
FeatureScatter(mydata_filtrd, feature1 = "nCount_RNA", feature2 = "percent.mt")
FeatureScatter(mydata_filtrd, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
```


# Vue d'ensemble d'une analyse de scRNA-seq

On rappelle qu'on a fait avant le filtrage, normalisation, et sélection des HVG.

```{r, eval = FALSE}
HVG <- Seurat::VariableFeatures(sobj)
HVG_plot <- Seurat::VariableFeaturePlot(sobj, selection.method = "sct")
Seurat::LabelPoints(plot = HVG_plot, points = head(HVG, 20), repel = TRUE)

head(sobj@assays$RNA@counts) #raw data
head(sobj@assays$SCT@counts) #raw data (there are copy paste into the new assay)
head(sobj@assays$SCT@data) #lognormalized data
head(sobj@assays$SCT@scale.data) #scal

DefaultAssay(sobj)

```


## Réduction de dimensions

```{r}
npcs <- 50
sobj <- RunPCA(sobj,
               features = VariableFeatures(object = sobj),
               verbose = FALSE,
               npcs = npcs)
```

```{r}
ElbowPlot(sobj, ndims = npcs)
```

```{r}
DimPlot(sobj,
		reduction = "pca",
		dims = c(1, 2),
		label = FALSE,
		label.size = 4,
		pt.size = 0.2)
```

## Clustering

```{r}
table(Idents(sobj))
```


```{r}
sobj <- FindNeighbors(sobj, dims = 1:npcs) %>%
    FindClusters(resolution = 0.3) %>%
	RunUMAP(dims = 1:npcs)
```

```{r}
table(Idents(sobj))
```

```{r}
DimPlot(sobj,
		reduction = "pca",
		dims = c(1, 2),
		label = FALSE,
		label.size = 4,
		pt.size = 0.2)
```

```{r}
sobj <- RunUMAP(sobj,
                dims = 1:npcs)
```

```{r}
DimPlot(sobj,
		reduction = "umap",
		dims = c(1, 2),
		label = FALSE,
		label.size = 4,
		pt.size = 0.2)
```

## Analyse d'expression différentielle

```{r}
markers_1 <- FindMarkers(sobj,
                         ident.1 = 1,
                         min.pct = 0.25)
```

```{r}
head(markers_1, n = 5)
```


```{r}
FeaturePlot(sobj,
            features = "Cd7",
            cols = c("grey90", rev(brewer.pal(5, "Spectral"))),
            pt.size = 0.2,
            ncol = 1) + NoLegend()
```
```{r}
VlnPlot(sobj, features = "Cd7")
```



```{r, message = FALSE}
markers_DE <- FindAllMarkers(sobj,
							 only.pos = TRUE,
							 min.pct = 0.25,
							 logfc.threshold = 0.25)

top10 <- markers_DE %>%
	filter(p_val < 1e-2) %>%
	group_by(cluster) %>%
	top_n(n = 10, wt = avg_log2FC)

DoHeatmap(sobj,
		  features = top10$gene,
		  label = FALSE) +
	theme(axis.text.y = element_text(size = 3))
```



# Réduction de dimensions avec la PCA
# Méthodes de clustering pour le single-cell

# Choix des clusters et annotations des cellules
# Analyse d'expression différentielle
```{r, message = FALSE}
markers_DE <- FindAllMarkers(sobj,
							 only.pos = TRUE,
							 min.pct = 0.25,
							 logfc.threshold = 0.25,
							 test.use = "wilcox")

markers_DE <- markers_DE %>% filter(p_val < 1e-2)

top10 <- markers_DE %>%
	filter(p_val < 1e-2) %>%
	group_by(cluster) %>%
	top_n(n = 10, wt = avg_log2FC)

DoHeatmap(sobj,
		  features = top10$gene,
		  label = FALSE) +
	theme(axis.text.y = element_text(size = 3))
```

# Pour aller plus loin

## Ressources {.unnumbered}

```{r, echo = FALSE, eval = FALSE}
This course material is largely inspired by two great sources that we advise you to read/use in details:

* [sincellTE 2020](https://cloud.biologie.ens.fr/index.php/s/RXkNGwtR1MLf5mo) : MA Dillies, A. Rausell, M Deloger, N. Servant, M. Thomas-Chollier, A. Paquet, cours6_TP_EDA_Normalization_2020.html
* [Vignette Seurat](https://satijalab.org/seurat/archive/v3.2/pbmc3k_tutorial.html)
```




